piker/brokers/kucoin.py

```diff
@@ -0,0 +1,668 @@
+# Copyright (C) Jared Goldman (in stewardship for pikers)
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+'''
+Kucoin broker backend
+
+'''
+
+from random import randint
+from typing import Any, Callable, Optional, Literal, AsyncGenerator
+from contextlib import asynccontextmanager as acm
+from datetime import datetime
+import time
+import math
+import base64
+import hmac
+import hashlib
+import wsproto
+from uuid import uuid4
+
+import asks
+import tractor
+import trio
+from trio_typing import TaskStatus
+from fuzzywuzzy import process as fuzzy
+import pendulum
+import numpy as np
+
+from piker._cacheables import open_cached_client
+from piker.log import get_logger
+from ._util import DataUnavailable
+from piker.pp import config
+from ..data.types import Struct
+from ..data._web_bs import (
+    open_autorecon_ws,
+    NoBsWs,
+)
+
+log = get_logger(__name__)
+
+_ohlc_dtype = [
```

Comment:
don't think you need to define this unless it's diff from our default.

Reply:
yah, but i think if you look, the default schema def is applied if not defined..

it should only be set if something is diff from the default iirc

we should probably document this.

Reply:
This confuses me. I'm using it to parse my data into the numpy array
```py
            for _, (field_name, field_type) in enumerate(_ohlc_dtype):
                value = data[field_name]

                match field_name:
                    case 'index':
                        row.append(int(value))
                    case 'time':
                        row.append(value)
                    case _:
                        row.append(float(value))

            new_bars.append(tuple(row))
```

Reply:
sure moreso my point is i don't think you need to define it at all in the module unless it's diff then the default:
https://github.com/pikers/piker/blob/master/piker/data/_source.py#L34

the logic which decides i believe is here in the data layer:
https://github.com/pikers/piker/blob/master/piker/data/feed.py#L799

so i'm just saying you might not need to even define it and if you need to iterate fields, just iterate the `base_iohlc_dtype`?

further this whole loop thinger:
```python
  for _, (field_name, field_type) in enumerate(_ohlc_dtype):
                value = data[field_name]
```

is this really any better then just doing inline static assignment?

for eg. here's how we do it in the `ib` backend:
https://github.com/pikers/piker/blob/master/piker/brokers/ib/api.py#L271

you'll notice it's just based on ordering the fields according to the client lib/API's output format, and we just kinda inline map it to the default ohlc dtype.

Reply:
Word, I'll simplify the logic